src/catalog/catalog.cpp

```diff
@@ -803,61 +801,155 @@ Catalog::~Catalog() {
 // FUNCTION
 //===--------------------------------------------------------------------===//
 
-void Catalog::AddFunction(
-    const std::string &name, const std::vector<type::TypeId> &argument_types,
-    const type::TypeId return_type,
-    type::Value (*func_ptr)(const std::vector<type::Value> &)) {
-  PL_ASSERT(functions_.count(name) == 0);
-  functions_[name] = FunctionData{name, argument_types, return_type, func_ptr};
+// add a new built-in function ("internal" language) into record
+// store the function in pg_proc as well as function::BuiltInFunctions
+// arguments:
+//    name & argument_types : function name and arg types used in SQL
+//    return_type : the return type
+//    prolang   : the oid of which language the function is
+//    func_name : the function name in C++ source code (should be unique)
+//    func_ptr  : the pointer to the function
+void Catalog::AddFunction(const std::string &name,
+                          const std::vector<type::TypeId> &argument_types,
+                          const type::TypeId return_type,
+                          oid_t prolang,
+                          const std::string &func_name,
+                          function::BuiltInFuncType func_ptr,
+                          codegen::function::BuiltInFuncType codegen_func_ptr_,
+                          concurrency::Transaction *txn) {
+  if (!ProcCatalog::GetInstance()->
+      InsertProc(name, return_type, argument_types,
+                 prolang, func_name, pool_.get(), txn)) {
+    throw CatalogException("Failed to add function " + func_name);
+  }
+  function::BuiltInFunctions::AddFunction(func_name, func_ptr);
+  codegen::function::BuiltInFunctions::AddFunction(func_name, codegen_func_ptr_);
 }
 
-const FunctionData Catalog::GetFunction(const std::string &name) {
-  if (functions_.count(name) == 0) {
-    throw Exception("function " + name + " not found.");
+const FunctionData Catalog::GetFunction(
+    const std::string &name,
+    const std::vector<type::TypeId> &argument_types) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  FunctionData result;
+  oid_t prolang = ProcCatalog::GetInstance()->GetProLang(name, argument_types, txn);
+
+  // if the function is "internal", then look up the map in function::BuiltInFunctions
+  // to get the function pointer
+  if (LanguageCatalog::GetInstance()->GetLanguageName(prolang, txn) == "internal") {
```

Comment:
Why are you doing lookup in the map only if the language is "internal"? It should be returning the function data even in the case of other languages.

Reply:
Yes, but I think function data should be returned in other ways if the function is in other language. The code only support "internal" language now, and we can add other languages and corresponding handle code in the future.